File: example/flux_example_executor/src/partitioned_sub.cpp
#include "flux/ros/partitioned_executor.hpp"
#include "flux/ros/subscription.hpp"

#include <rclcpp/rclcpp.hpp>

#include "sensor_msgs/flux/image.hpp"

#include <chrono>
#include <cstdint>
#include <memory>
#include <thread>

// Two flux subscriptions, two callback groups, two threads. A slow callback in one group cannot
// delay the other -- which is the whole reason to pay for the extra thread.
class PartitionedSubscriber : public rclcpp::Node
{
public:
  PartitionedSubscriber()
  : Node("flux_partitioned_sub"),
    fast_group_(create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive)),
    slow_group_(create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive)),
    fast_(*this, kTopic, Image::kFingerprint, [this](const flux::FrameView & f) { on_fast(f); }),
    slow_(*this, kTopic, Image::kFingerprint, [this](const flux::FrameView & f) { on_slow(f); })
  {
    timer_ = create_wall_timer(std::chrono::seconds(1), [this]() { report(); });
  }

  rclcpp::CallbackGroup::SharedPtr fast_group() { return fast_group_; }
  rclcpp::CallbackGroup::SharedPtr slow_group() { return slow_group_; }
  flux::ros::Subscription & fast() { return fast_; }
  flux::ros::Subscription & slow() { return slow_; }

private:
  using Image = sensor_msgs::flux_msg::Image;

  static constexpr const char * kTopic = "image";
  static constexpr std::chrono::milliseconds kSlowWork{80};

  void on_fast(const flux::FrameView & f)
  {
    Image::View v{f};
    if (v.ok__()) {
      ++fast_seen_;
    }
  }

  void on_slow(const flux::FrameView & f)
  {
    Image::View v{f};
    if (v.ok__()) {
      std::this_thread::sleep_for(kSlowWork);
      ++slow_seen_;
    }
  }

  void report() { RCLCPP_INFO(get_logger(), "fast %lu  slow %lu", fast_seen_, slow_seen_); }

  rclcpp::CallbackGroup::SharedPtr fast_group_;
  rclcpp::CallbackGroup::SharedPtr slow_group_;
  flux::ros::Subscription fast_;
  flux::ros::Subscription slow_;
  rclcpp::TimerBase::SharedPtr timer_;
  std::uint64_t fast_seen_ = 0;
  std::uint64_t slow_seen_ = 0;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<PartitionedSubscriber>();

  flux::ros::PartitionedExecutor ex;
  ex.add_ros_node(node);
  ex.add(node->fast(), node->fast_group());
  ex.add(node->slow(), node->slow_group());

  rclcpp::on_shutdown([&ex]() { ex.stop(); });
  ex.spin();
  rclcpp::shutdown();
  return 0;
}

File: example/flux_example_executor/flux_example_executor/partitioned_sub.py
#!/usr/bin/env python3
"""Give each flux subscription its own thread, so a slow callback delays nobody else.

ros2 run flux_example_executor partitioned_sub.py
"""

import time

import flux.ros
import rclpy
from rclpy.callback_groups import MutuallyExclusiveCallbackGroup
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from sensor_msgs_flux.image import Image


class PartitionedSubscriber(Node):

    TOPIC = "image"
    SLOW_WORK_SEC = 0.08

    def __init__(self):
        super().__init__("flux_partitioned_sub")
        self.fast_seen = 0
        self.slow_seen = 0

        # Partition tokens only. rclpy cannot move a ROS entity onto a child executor's thread,
        # so a group handed to add_flux must hold none.
        self.fast_group = MutuallyExclusiveCallbackGroup()
        self.slow_group = MutuallyExclusiveCallbackGroup()

        self.fast_sub = flux.ros.Subscription(
            self, self.TOPIC, callback=self.on_fast, fingerprint=Image.FINGERPRINT__
        )
        self.slow_sub = flux.ros.Subscription(
            self, self.TOPIC, callback=self.on_slow, fingerprint=Image.FINGERPRINT__
        )

        self.timer = self.create_timer(1.0, self.report)

    def on_fast(self, _view):
        self.fast_seen += 1

    def on_slow(self, _view):
        time.sleep(self.SLOW_WORK_SEC)
        self.slow_seen += 1

    def report(self):
        self.get_logger().info(f"fast {self.fast_seen}  slow {self.slow_seen}")


def main(args=None):
    rclpy.init(args=args)
    node = PartitionedSubscriber()
    executor = flux.ros.PartitionedExecutor()
    executor.add_ros_node(node)
    executor.add_flux(node.fast_sub, node.fast_group)
    executor.add_flux(node.slow_sub, node.slow_group)
    try:
        executor.spin()
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        executor.stop()
        executor.close()
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == "__main__":
    main()
